linera-views/src/memory.rs

```diff
@@ -193,84 +201,109 @@ impl WritableKeyValueStore<MemoryStoreError> for MemoryStore {
-        namespace_memory_store: &NamespaceMemoryStore,
+        memory_stores: &mut MemoryStores,
         config: &MemoryStoreConfig,
         namespace: &str,
+        root_key: &[u8],
         kill_on_drop: bool,
     ) -> Result<Self, MemoryStoreError> {
         let max_stream_queries = config.common_config.max_stream_queries;
-        let namespace = namespace.to_string();
-        let store = namespace_memory_store
-            .get(&namespace)
-            .ok_or(MemoryStoreError::NotExistentNamespace)?;
+        ensure!(
+            memory_stores.namespaces.contains(namespace),
+            MemoryStoreError::NotExistentNamespace
+        );
+        let pair = (namespace.to_string(), root_key.to_vec());
+        let store = memory_stores.stores.entry(pair).or_insert_with(|| {
+            let map = MemoryStoreMap::new();
+            Arc::new(RwLock::new(map))
+        });
         let map = store.clone();
         let namespace = namespace.to_string();
+        let root_key = root_key.to_vec();
         Ok(MemoryStore {
             map,
             max_stream_queries,
             namespace,
+            root_key,
             kill_on_drop,
         })
     }
 
-    fn sync_list_all(namespace_memory_store: &NamespaceMemoryStore) -> Vec<String> {
-        namespace_memory_store.keys().cloned().collect::<Vec<_>>()
+    fn sync_list_all(memory_stores: &MemoryStores) -> Vec<String> {
+        memory_stores.namespaces.iter().cloned().collect::<Vec<_>>()
     }
 
-    fn sync_exists(namespace_memory_store: &NamespaceMemoryStore, namespace: &str) -> bool {
-        let namespace = namespace.to_string();
-        namespace_memory_store.contains_key(&namespace)
+    fn sync_exists(memory_stores: &MemoryStores, namespace: &str) -> bool {
+        memory_stores.namespaces.contains(namespace)
     }
 
-    fn sync_create(namespace_memory_store: &mut NamespaceMemoryStore, namespace: &str) {
-        let namespace = namespace.to_string();
-        let map = MemoryStoreMap::new();
-        let map = Arc::new(RwLock::new(map));
-        namespace_memory_store.insert(namespace, map);
+    fn sync_create(memory_stores: &mut MemoryStores, namespace: &str) {
+        memory_stores.namespaces.insert(namespace.to_string());
     }
 
-    fn sync_delete(namespace_memory_store: &mut NamespaceMemoryStore, namespace: &str) {
+    fn sync_delete(memory_stores: &mut MemoryStores, namespace: &str) {
         let namespace = namespace.to_string();
-        namespace_memory_store.remove(&namespace);
+        memory_stores.namespaces.remove(&namespace);
+        let mut pair_removes = Vec::new();
+        for key in memory_stores.stores.keys() {
+            if key.0 == namespace {
+                pair_removes.push(key.clone());
+            }
+        }
+        for pair in pair_removes {
+            memory_stores.stores.remove(&pair);
+        }
     }
 
     /// Create a memory store if one is missing and otherwise connect with the existing one
```

Comment:
```suggestion
    /// Creates a memory store if one is missing and otherwise connects with the existing one
```

Or maybe: "Connects to a memory store. Creates it if it doesn't exist yet.

Reply:
Ok.